src/Microsoft.Extensions.Configuration.AzureAppConfiguration/DomainVerifier.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright(c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration
+{
+    /// <summary>
+    /// Verify the domain name of the endpoint matches the certificate.
+    /// </summary>
+    internal static class DomainVerifier
+    {
+        private const int TlsPort = 443;
+        private const string SubjectAltNameOid = "2.5.29.17";
+        private const string DnsName = "DNS Name=";
+
+        public static async Task<IEnumerable<string>> GetValidDomain(Uri originEndpoint, string srvHostName)
+        {
+            IEnumerable<string> validDomains = await GetSubjectAlternativeNames(srvHostName).ConfigureAwait(false);
+
+            var domainList = new List<string>();
+            
+            foreach (string domain in validDomains)
+            {
+                if (originEndpoint.DnsSafeHost.EndsWith(domain, StringComparison.OrdinalIgnoreCase))
+                {
+                    domainList.Add(domain);
+                }
+            }
+
+            return domainList;
+        }
+
+        private static async Task<IEnumerable<string>> GetSubjectAlternativeNames(string endpoint)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(endpoint));
+
+            var validDomains = new List<string>();
+
+            X509Certificate2 cert = null;
+
+            using (var client = new TcpClient(endpoint, TlsPort))
+            using (var sslStream = new SslStream(client.GetStream(), leaveInnerStreamOpen: false))
+            {
+                // Initiate the connection, so it will download the server certificate
+                await sslStream.AuthenticateAsClientAsync(endpoint).ConfigureAwait(false);
+
+                X509Certificate serverCertificate = sslStream.RemoteCertificate;
+
+                if (serverCertificate == null)
+                {
+                    return validDomains;
+                }
+
+                cert = new X509Certificate2(serverCertificate);
+            }
+
+            if (cert == null)
+            {
+                return validDomains;
+            }
+
+            using (cert)

```

Comment:
It's a bit unfortunate the `using` is separated from the insantiation of the disposable object.

Could you use a private method to avoid separating the two?

```
using (X509Certificate2 cert = new X509Certificate2(serverCertificate))
{
    return GetValidDomains(cert);
}
```